src/wireviz/wireviz.py

```diff
@@ -90,108 +58,92 @@ def check_designators(what, where):
+        return True
+
+    autogenerated_ids = {}
+    for connection in yaml_data['connections']:
+        # TODO: check that items are of alternating type CONNECTOR/FERRULE/FERRULE_LIST and CABLE/WIRE
+        # TODO: special case: loops!
+
+        # check that all iterable items (lists and dicts) are the same length
+        itemcount = None
+        for item in connection:
+            if isinstance(item, list):
+                itemcount_new = len(item)
+            elif isinstance(item, dict):
+                if len(item.keys()) != 1:
+                    raise Exception('Dicts may contain only one item here!')
+                itemcount_new = len(expand(list(item.values())[0]))
+            elif isinstance(item, str):
+                continue
+            if itemcount is not None and itemcount_new != itemcount:
+                raise Exception('All lists and dict lists must be the same length!')
+            itemcount = itemcount_new
+        if itemcount is None:
+            raise Exception('No item revealed the number of connections to make!')
+
+        # populate connection list
+        connection_list = []
+        for i, item in enumerate(connection):
+            if isinstance(item, str):  # one single-pin component was specified
+                sublist = []
+                for i in range(1, itemcount + 1):
+                    if yaml_data['connectors'][item].get('autogenerate'):
+                        autogenerated_ids[item] = autogenerated_ids.get(item, 0) + 1
+                        new_id = f'_{item}_{autogenerated_ids[item]}'
+                        harness.add_connector(new_id, **yaml_data['connectors'][item])
+                        sublist.append([new_id, 1])
+                    else:
+                        sublist.append([item, 1])
+                connection_list.append(sublist)
+            elif isinstance(item, list):  # a list of single-pin components were specified
+                sublist = []
+                for subitem in item:
+                    if yaml_data['connectors'][subitem].get('autogenerate'):
+                        autogenerated_ids[subitem] = autogenerated_ids.get(subitem, 0) + 1
+                        new_id = f'_{subitem}_{autogenerated_ids[subitem]}'
+                        harness.add_connector(new_id, **yaml_data['connectors'][subitem])
+                        sublist.append([new_id, 1])
+                    else:
+                        sublist.append([subitem, 1])
+                connection_list.append(sublist)
+            elif isinstance(item, dict):  # a component with multiple pins was specified
+                sublist = []
+                id = list(item.keys())[0]
+                pins = expand(list(item.values())[0])
+                for pin in pins:
+                    sublist.append([id, pin])
+                connection_list.append(sublist)
+            elif False:  # TODO: placeholer; a loop inside a connector was specified
+                pass
```

Comment:
Ideas?

Reply:
Have you considered this?
`connection_list.append([(id,pin) for pin in pins])`
If the (id,pin) pair is never modified, then I suggest storing it as a tuple instead of a list.

Reply:
```
def conn_item(item):
    if not yaml_data['connectors'][item].get('autogenerate'):
        return item
    autogenerated_ids[item] = autogenerated_ids.get(item, 0) + 1
    new_id = f'_{item}_{autogenerated_ids[item]}'
    harness.add_connector(new_id, **yaml_data['connectors'][item])
    return new_id
```

Reply:
That won't work. I'll re-implement loops elsewhere, separate from regular connection sets.